### ISSUE TITLE
[Bounty $2,500] binary_ng/ternary: replicate the broadcast row with local NoC copies instead of ~2000 scalar stores per tile

### ISSUE THREAD

**Issue by kinginu** [tenstorrent/tt-metal#55138]
**Component / Area:** ttnn eltwise dataflow kernels (`fill_tile_utils.hpp`, binary_ng / ternary / quasar readers)

**Issue Type:** Performance

## Observed

`fill_tile_with_first_row` replicates one row across a tile with a scalar loop on a data-movement core, re-reading its source through a `volatile` pointer on every iteration of the innermost loop:

```cpp
for (uint32_t k = 0, face_offset = 0; k < 2; ++k, face_offset += 512) {
    for (uint32_t row = 0; row < num_rows; ++row) {
        for (uint32_t col = 0; col < 16; ++col) {
            ptr[dst_offset + col]       = ptr[col];        // volatile load, every iteration
            ptr[dst_offset + col + 256] = ptr[col + 256];
        }
        row_offset += 16;
    }
}
```

That is ~992 volatile loads and ~992 stores per tile, and the row readers call it **once per output tile** — about two million of each for a 1024-tile call. The column reader next to it hoists its fill out of the inner loop and hoists its own load, and measures at 0.78-1.03x; only the row path pays.

## Expected

Within a face the 16 rows are contiguous — 16 words, 64 B per row, 1024 B per face — so doubling the valid region fills a face in four transfers. Faces 0 and 1 are the top half of the tile and hold the source row; faces 2 and 3 are the bottom half and have no row of their own, which is why the scalar version fills 15 rows for the first face pair and 16 for the second. Double within the top faces, then copy each finished face into the one below it.

```cpp
FORCE_INLINE void fill_tile_with_first_row(uint32_t l1_write_ptr) {
    constexpr uint32_t kFaceBytes = 1024;
    constexpr uint32_t kRowBytes = 64;
    const uint32_t f0 = l1_write_ptr;
    const uint32_t f1 = f0 + kFaceBytes;
    for (uint32_t n = kRowBytes; n < kFaceBytes; n <<= 1) {
        noc_async_read(get_noc_addr(f0), f0 + n, n);
        noc_async_read(get_noc_addr(f1), f1 + n, n);
        noc_async_read_barrier();
    }
    noc_async_read(get_noc_addr(f0), f0 + 2 * kFaceBytes, kFaceBytes);
    noc_async_read(get_noc_addr(f1), f1 + 2 * kFaceBytes, kFaceBytes);
    noc_async_read_barrier();
}
```

Ten transfers and five barriers per tile against ~1,984 volatile memory operations. #52909 makes the same move for `generate_index_tile`, and `writer_deepseek_grouped_gate.cpp:30` already uses the construction in-tree.

## Measured

Both machines at `047ec3ccf96`, per-op wall clock in microseconds, median of 7 runs of 30 back-to-back calls with a `synchronize_device` per run. Implemented and measured, not proposed. An ablation that deletes the replication entirely — wrong results by construction, run only to bound the saving — lands within noise of the after column at every large shape, so the fill is essentially the whole gap. Tables are per machine because the two behave differently and for opposite reasons.

### Blackhole P300

| op | shape | dtype | broadcast operands | before | after | |
|---|---|---|---|---|---|---|
| `ttnn.where` | `[1,1,1024,1024]` | f32 | 2 rows | 226.24 | **43.15** | -80.9% |
| `ttnn.where` | `[1,1,32,32768]` | f32 | 2 rows | 226.99 | **44.27** | -80.5% |
| `ttnn.where` | `[1,1,32,8192]` | f32 | 2 rows | 74.32 | **16.50** | -77.8% |
| `ttnn.multiply` | `[1,1,1024,1024]` | f32 | 1 row | 118.06 | **31.93** | -73.0% |
| `ttnn.add` | `[1,1,32,32768]` | f32 | 1 row | 117.91 | **31.96** | -72.9% |
| `ttnn.multiply` | `[1,1,32,32768]` | f32 | 1 row | 117.94 | **31.98** | -72.9% |
| `ttnn.gt` | `[1,1,32,32768]` | f32 | 1 row | 118.05 | **32.02** | -72.9% |
| `ttnn.add` | `[1,1,1024,1024]` | f32 | 1 row | 118.13 | **32.06** | -72.9% |
| `ttnn.gt` | `[1,1,1024,1024]` | f32 | 1 row | 118.30 | **32.30** | -72.7% |
| `ttnn.where` | `[1,1,32,2048]` | f32 | 2 rows | 26.07 | **8.77** | -66.4% |
| `ttnn.where` | `[1,1,1024,1024]` | f32 | 1 row | 120.59 | **42.34** | -64.9% |
| `ttnn.where` | `[1,1,32,32768]` | f32 | 1 row | 121.40 | **43.44** | -64.2% |
| `ttnn.where` | `[1,1,32,8192]` | f32 | 1 row | 42.70 | **15.36** | -64.0% |
| `ttnn.where` | `[1,1,32,2048]` | f32 | 1 row | 15.53 | **8.84** | -43.1% |
| `ttnn.gt` | `[1,1,32,8192]` | f32 | 1 row | 40.36 | **25.00** | -38.1% |
| `ttnn.multiply` | `[1,1,32,8192]` | f32 | 1 row | 40.17 | **25.11** | -37.5% |
| `ttnn.add` | `[1,1,32,8192]` | f32 | 1 row | 40.17 | **25.34** | -36.9% |
| `ttnn.where` | `[1,1,32,2048]` | bf16 | 2 rows | 8.83 | **8.49** | -3.9% |
| `ttnn.gt` | `[1,1,32,2048]` | f32 | 1 row | 25.35 | **24.78** | -2.2% |
| `ttnn.gt` | `[1,1,1024,1024]` | bf16 | 1 row | 28.65 | **28.03** | -2.2% |
| `ttnn.gt` | `[1,1,32,8192]` | bf16 | 1 row | 25.14 | **24.62** | -2.1% |

The remaining 19 of 40 measured rows do not move beyond the 2% noise floor on this machine.

### Wormhole N300

| op | shape | dtype | broadcast operands | before | after | |
|---|---|---|---|---|---|---|
| `ttnn.where` | `[1,1,32,32768]` | f32 | 2 rows | 547.81 | **122.30** | -77.7% |
| `ttnn.where` | `[1,1,1024,1024]` | f32 | 2 rows | 551.91 | **134.14** | -75.7% |
| `ttnn.where` | `[1,1,32,8192]` | f32 | 2 rows | 166.42 | **62.17** | -62.6% |
| `ttnn.where` | `[1,1,32,32768]` | f32 | 1 row | 304.83 | **117.42** | -61.5% |
| `ttnn.where` | `[1,1,1024,1024]` | f32 | 1 row | 307.21 | **124.50** | -59.5% |
| `ttnn.where` | `[1,1,32,2048]` | f32 | 2 rows | 49.08 | **26.24** | -46.5% |
| `ttnn.where` | `[1,1,32,8192]` | f32 | 1 row | 103.72 | **62.41** | -39.8% |
| `ttnn.where` | `[1,1,32,8192]` | bf16 | 1 row | 38.57 | **24.88** | -35.5% |
| `ttnn.where` | `[1,1,32,8192]` | bf16 | 2 rows | 32.28 | **24.89** | -22.9% |

The remaining 31 of 40 measured rows do not move beyond the 10% noise floor on this machine.

Baselines reproduce to under 2% on P300. **On N300 they do not**: the small shapes there are unstable, with run-to-run spreads from 6% to 118% on repeated sampling, so that machine's cut is drawn at 10% and every row inside it is reported as unmoved rather than as a small change. One row initially read as a +58% regression and was a sampling artifact — re-measured with 15 runs of 50 calls it is 24.37 against 24.38.

Which rows move is decided by whether that combination already had a hardware broadcast:

- **bfloat16 does not move, and does not need to.** `unary_bcast<BroadcastType::ROW>` lowers to `MOVD2B` / `MOVB2D` with `MOV_8_ROW_BRCST`, run from a MOP: the replication happens inside the Dst register file, eight instructions for a tile, and the reader's fill is compiled out entirely (`#if SRC_BCAST && !BCAST_LLK`). There is nothing here to improve.
- **float32 has no such path in the ternary readers.** `ternary_program_factory.cpp:218` admits the hardware broadcast only when all three operands are bfloat16. Forcing float32 through it anyway returns wrong values for every element, on **both** architectures — not the Blackhole-specific tenstorrent/tt-llk#449, but because the 32-bit two-pass sequence exists only in the `A2D` unpack-to-dest branch and not in the `B2D` + MOP branch that `unary_bcast` uses. So the bfloat16-only condition is correct, and software replication is the only option there.
- **float32 in binary_ng moves on Blackhole only**, because `is_llk_bcast` already routes the 32-bit dtypes through the hardware broadcast on Wormhole. That exclusion is a separate report.

## Correctness

Outputs bit-identical to the scalar fill at every shape and dtype measured, on both machines. `ttnn.where` with row-broadcast operands checked against `torch.where` at 3 widths, 0 of 65,536 elements wrong.

## Scope

| site | |
|---|---|
| `eltwise/binary_ng/device/kernels_ng/dataflow/reader_interleaved_row_bcast.cpp` | in scope |
| `eltwise/binary_ng/.../reader_interleaved_row_col_mixed_bcast.cpp`, `_scalar_bcast.cpp` | in scope |
| `eltwise/ternary/.../ternary_reader_rowbcast_ttt.cpp` | in scope, fills **three** operands per output tile |
| `eltwise/ternary/.../tts_tst_reader_row_col_bcast.cpp` | in scope |
| `experimental/quasar/binary_ng/...` | a separate duplicate of the same function, **not** changed here |
| `fill_tile_with_first_column` | already hoisted, measures 0.78-1.03x, **not** in scope |
| bfp8 / bfp4 fills | already packed, not per-element, not in scope |

Transposing a column broadcast into a row broadcast is not worth it: the column path measures 0.78-0.90x, i.e. **faster** than a same-shape operand, because a column operand is one tile per tile row and the read saving outweighs the fill.

## Not established

- A mechanism that would remove the replication altogether rather than make it cheaper: the packer can source datums from L1, and a zero row stride would pack every output row from the same source row. Unmeasured, and an API-level change, so it is not proposed here.
- Only float32 and bfloat16 were timed, through `multiply` / `add` / `gt` / `where`.
- The quasar readers were read, not measured.
- `noc_async_write` in place of the loop-back read measures identically (43.67 vs 43.72 us), so the direction of the local transfer does not matter; the read form is kept because it matches the existing in-tree precedent.
- The barrier per doubling level waits on all outstanding reads on that NoC. The current readers already barrier per tile so nothing is lost, but a reader that prefetched the next tile would lose that overlap.

**Comment by KalaivaniMCW** [comment]
@kinginu 
`fill_tile_with_first_row` is only for 32-bit dtypes and wouldn't affect 16-bit dtypes 
Did you observe a change in your testing ? 
What are the before & after  referring to in the table? an in what metrics?



<br class="Apple-interchange-newline">

**Comment by kinginu** [comment]
Thanks — you are right on both counts, and the bf16 rows in the table were wrong. Details:

**1. Scope.** The change is the 32-bit `fill_tile_with_first_row` only. bfloat16 dispatches to a separate function, `fill_tile_with_first_row_bfloat16` (`binary_ng_utils.cpp` / `ternary_op_utils.cpp` select it by dtype), which the patch never touched. So nothing in that patch could have moved a bf16 row, and the bf16 deltas in the table were measurement noise: on N300 the bf16 binary_ng rows vary by ±10-20% between two runs of the *unchanged* tree (e.g. `add [1,1,1024,1024]` 51.2 vs 61.3 µs), which is the size of the "gains" I reported. I am withdrawing the bf16 rows.

To settle it rather than argue it, I re-measured today on the 2026-09-02 nightly (`2aa657a22e9`) on both machines, with the bf16 variant **also** converted to the same NoC doubling (32 B rows, 512 B faces, 4 doublings + 2 face copies) so that any bf16 effect would show:

| dtype | P300 (Blackhole) | N300 (Wormhole) |
|---|---|---|
| bf16, all 20 rows (`where` 1 and 2 broadcast operands, `multiply`/`add`/`gt`, four shapes) | within ±3% | within run-to-run variance |

No bf16 row moves on either machine even with its fill converted. The bf16 fill is simply not on the critical path: binary_ng bf16 broadcasts take the LLK hardware path on both architectures, and `where` bf16 sits at the dispatch floor (22 µs at 1024 tiles, 9 µs at 256 tiles on P300).

**2. Metric.** Per-op wall clock in microseconds: 5 warm-ups, then 30 back-to-back calls with one `synchronize_device`, repeated 7 times, median. "Before" is the tree as shipped; "after" is the same tree with `fill_tile_with_first_row` replaced by the NoC copies (a kernel-only change, same build). Outputs are bit-identical (sha256 over the raw output bits) in every cell.

The 32-bit rows reproduce on the nightly, same method, same day:

| op, shape, float32 | P300 before → after | | N300 before → after | |
|---|---|---|---|---|
| `where` [1,1,1024,1024], 2 row operands | 226.6 → **42.6** | −81% | 546.6 → **133.7** | −76% |
| `where` [1,1,1024,1024], 1 row operand | 120.7 → **42.6** | −65% | 307.2 → **115.5** | −62% |
| `where` [1,1,32,32768], 2 / 1 | 227.1 → 43.9 / 121.3 → 42.9 | −81% / −65% | 547.4 → 122.3 / 304.7 → 120.0 | −78% / −61% |
| `where` [1,1,32,8192], 2 / 1 | 74.4 → 16.5 / 42.8 → 15.8 | −78% / −63% | 165.9 → 62.3 / 103.9 → 53.3 | −62% / −49% |
| `multiply` / `add` / `gt` [1,1,1024,1024], 1 row operand | 118.2 → **32.3** | −73% | 96.9 → 96.9 | 0 (LLK broadcast path on Wormhole; the fill does not run) |
| `multiply` / `add` / `gt` [1,1,32,8192] | 40.5 → 25.2 | −38% | ~50 → ~50 | 0 (same) |
| `multiply` / `add` / `gt` [1,1,32,2048] | 24.8 → 24.5 | 0 (dispatch floor) | ~50 → ~50 | 0 |

So the claim as it should have been stated: **float32 / int32 / uint32 row broadcasts pay for the scalar fill — `where` on both architectures, binary_ng on Blackhole (where the 32-bit dtypes are excluded from the LLK broadcast path) — and the NoC replication removes 60-80% of the op's time at ≥256 tiles.** The PR will change the 32-bit function; converting the bf16 one is harmless and symmetric but has no measurable effect, so I will leave it out unless you prefer the two to match.

**Comment by jberkowitzTT** [comment]
## 📝 Background

This is a community-suggested bounty that we have reviewed & assigned to @kinginu 


## 💡 Problem to Solve

>binary_ng/ternary: replicate the broadcast row with local NoC copies instead of ~2000 scalar stores per tile


Note: This bounty is beholden to the [Tenstorrent Bounty Program Terms & Conditions](https://docs.tenstorrent.com/bounty_terms.html) & local tax requirements may apply

**Comment by jberkowitzTT** [comment]
Hi @kinginu - congratulations on completing this bounty, and thank you for the great work!

To initiate the payment process, please email us at bounties@tenstorrent.com with a link to this issue
